### PR TITLE
Add SubscriptionType entity for email notification

### PR DESCRIPTION
**Pull request by ottaviano**
https://app.clubhouse.io/enmarche/story/4294/step-1-refaire-le-squelette-des-notifications

image

### CODE REVIEW

File: src/DataFixtures/ORM/LoadEmailSubscriptionHistoryData.php [hunk 1]
```diff
@@ -76,27 +76,22 @@ public function load(ObjectManager $manager)
         $manager->flush();
     }
 
-    private function createSubscribedUnsubscribedHistory(
-        Adherent $adherent,
-        array $tags,
-        string $subscribedAt,
-        string $unsubscribedAt = '-1 month',
```
Comment (illusionOfParadise) [inline src/DataFixtures/ORM/LoadEmailSubscriptionHistoryData.php]: This argument is still passed to the method https://github.com/EnMarche/en-marche.fr/pull/2546/files#diff-c8b7e31a004d47136afcbde7cb6b1e82R61, but no more taken into account.

File: app/migrations/Version20180620154117.php [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        $this->addSql('ALTER TABLE adherent_subscription_type ADD CONSTRAINT FK_F93DC28A25F06C53 FOREIGN KEY (adherent_id) REFERENCES adherents (id) ON DELETE CASCADE');
+        $this->addSql('ALTER TABLE adherent_subscription_type ADD CONSTRAINT FK_F93DC28AB6596C08 FOREIGN KEY (subscription_type_id) REFERENCES subscription_type (id) ON DELETE CASCADE');
+        $this->addSql('ALTER TABLE adherents DROP local_host_emails_subscription, DROP emails_subscriptions');
+    }
+
+    public function postUp(Schema $schema)
+    {
+        $this->connection->executeQuery(
+            "INSERT INTO subscription_type (label,code) VALUES 
+('Emails de votre animateur local','subscribed_emails_local_host')
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply app/migrations/Version20180620154117.php]: @illusionOfParadise, fixed 👍

File: app/migrations/Version20180620172122.php [hunk 3]
```diff
@@ -0,0 +1,41 @@
+<?php
+
+namespace Migrations;
+
+use Doctrine\DBAL\Migrations\AbstractMigration;
+use Doctrine\DBAL\Schema\Schema;
+
+final class Version20180620172122 extends AbstractMigration
```
[the remark this answers is not in the document]
Comment (ottaviano) [reply app/migrations/Version20180620172122.php]: @illusionOfParadise, fixed

File: src/DataFixtures/ORM/LoadEmailSubscriptionHistoryData.php (hunk 1 above)
Comment (ottaviano) [reply src/DataFixtures/ORM/LoadEmailSubscriptionHistoryData.php]: @illusionOfParadise, the 4-th and 5-th arguments are never passed, that is why I remove them

File: app/config/services/services.xml [hunk 4]
```diff
@@ -226,5 +226,9 @@
         </service>
 
         <service id="AppBundle\Security\ApiAuthenticationEntryPoint"/>
+
+        <!-- Subscription -->
+        <service id="AppBundle\Subscription\SubscriptionListener" />
```
Comment (HeahDude) [inline app/config/services/services.xml]: Why not registering the namespace?

File: src/Command/MigrateSubscriptionTypeCommand.php [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        $this->setDescription('Migrate the subscription types from adherent.emails_subscriptions column to new table: subscription_type');
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+        foreach ($this->getIterator() as $index => $adherent) {
+            /** @var Adherent $adherent */
+            $adherent = $adherent[0];
+
+            $adherentTypes = array_filter($adherent->emailsSubscriptions, function (string $type) {
```
Comment (HeahDude) [inline src/Command/MigrateSubscriptionTypeCommand.php]: missing `\`?

File: src/Entity/Adherent.php [hunk 6]
```diff
@@ -229,6 +220,23 @@ class Adherent implements UserInterface, GeoPointInterface, EncoderAwareInterfac
      */
     private $roles = [];
 
+    /**
+     * @ORM\Column(type="boolean")
```
Comment (HeahDude) [inline src/Entity/Adherent.php]: The column is not nullable b y default though the default should also be set at db level through the mapping:
```php
/**
 * @ORM\Column(type="boolean", options={"default"=false})
 */
```
and the migration will need to be updated.

File: src/Entity/SubscriptionType.php [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+ * @ORM\Table(indexes={@ORM\Index(columns={"code"})})
+ *
+ * @UniqueEntity(fields={"code"})
+ *
+ * @Algolia\Index(autoIndex=false)
+ */
+class SubscriptionType
+{
+    /**
+     * @var int
```
Comment (HeahDude) [inline src/Entity/SubscriptionType.php]: `|null`

File: src/Entity/SubscriptionType.php [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+     *
+     * @ORM\Column(type="integer", options={"unsigned": true})
+     * @ORM\Id
+     * @ORM\GeneratedValue
+     */
+    private $id;
+
+    /**
+     * @var string
```
Comment (HeahDude) [inline src/Entity/SubscriptionType.php]: `|null` (same below)

File: src/Entity/SubscriptionType.php [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    {
+        return $this->id;
+    }
+
+    public function getLabel(): ?string
+    {
+        return $this->label;
+    }
+
+    public function setLabel(string $label): void
```
Comment (HeahDude) [inline src/Entity/SubscriptionType.php]: Why not keeping it nullable here?

File: src/Form/AdherentEmailSubscriptionType.php [hunk 10]
```diff
@@ -43,15 +65,25 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
                 ]);
-                $formData['citizenProjectCreationEmailSubscriptionRadius'] = -1;
-                $event->setData($formData);
+                $data['citizenProjectCreationEmailSubscriptionRadius'] = Adherent::DISABLED_CITIZEN_PROJECT_EMAIL;
+                $event->setData($data);
             }
         });
     }
+
+    public function configureOptions(OptionsResolver $resolver)
```
Comment (HeahDude) [inline src/Form/AdherentEmailSubscriptionType.php]: You should move this method above for readability (we'll change that soon in the docs).

File: src/Form/AdherentEmailSubscriptionType.php [hunk 11]
```diff
@@ -43,15 +65,25 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
+    {
+        $resolver->setDefaults([
+            'is_adherent' => true,
+        ]);
+
+        $resolver->setAllowedTypes('is_adherent', 'bool');
```
Comment (HeahDude) [inline src/Form/AdherentEmailSubscriptionType.php]: Method calls can be chained, the resolver uses a builder pattern

File: src/Repository/SubscriptionTypeRepository.php [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+use Symfony\Bridge\Doctrine\RegistryInterface;
+
+class SubscriptionTypeRepository extends ServiceEntityRepository
+{
+    public function __construct(RegistryInterface $registry)
+    {
+        parent::__construct($registry, SubscriptionType::class);
+    }
+
+    public function findByCodes(array $codes): array
```
Comment (HeahDude) [inline src/Repository/SubscriptionTypeRepository.php]: You should add:
```php
/**
 * @param int[]
 *
 * @return SubscriptionType[]
 */
```
to help autocompletion.

File: src/Entity/Adherent.php [hunk 13]
```diff
@@ -1004,29 +957,38 @@ public function setLegislativeCandidate(bool $candidate): void
...
+    public function removeSubscriptionType(SubscriptionType $type): void
     {
-        $this->setCitizenProjectCreationEmailSubscriptionRadius(
-            $comEmail ? self::CITIZEN_PROJECT_EMAIL_DEFAULT_DISTANCE : self::DISABLED_CITIZEN_PROJECT_EMAIL
-        );
+        $this->subscriptionTypes->removeElement($type);
+    }
 
-        $this->localHostEmailsSubscription = $comEmail;
+    public function removeSubscriptionTypeByCode(string $code): void
```
Comment (HeahDude) [inline src/Entity/Adherent.php]: When is this method used?

What about simply:
```php
foreach ($this->subscriptionTypes as $type) {
     if ($code === $type->getCode()) {
        $this->removeSubscriptionType($type);
    }
}
```
Not sure we need to introduce Doctrine related complexity to the model.

**Pull request merged**